runtime/vm/ContinuationHelpers.cpp

```diff
@@ -0,0 +1,198 @@
+		UDATA usedBytes = ((UDATA)newELS.oldEntryLocalStorage - (UDATA)&newELS);
+		freeBytes -= usedBytes;
+		currentThread->currentOSStackFree = freeBytes;
+
+		if ((IDATA)freeBytes < J9_OS_STACK_GUARD) {
+			if (J9_ARE_NO_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_CONSTRUCTING_EXCEPTION)) {
+				setCurrentExceptionNLS(currentThread, J9VMCONSTANTPOOL_JAVALANGSTACKOVERFLOWERROR, J9NLS_VM_OS_STACK_OVERFLOW);
+				currentThread->currentOSStackFree += usedBytes;
+			}
+		}
+	}
+
+	VM_OutOfLineINL_Helpers::restoreInternalNativeStackFrame(currentThread);
+	VM_OutOfLineINL_Helpers::returnSingle(currentThread, JNI_TRUE, 1);
+
+	currentThread->returnValue = J9_BCLOOP_EXECUTE_BYTECODE;
+
+	c_cInterpreter(currentThread);
+
+	restoreCallInFrame(currentThread);
+}
+
+BOOLEAN
+enterContinuation(J9VMThread *currentThread, j9object_t continuationObject)
+{
+	BOOLEAN result = TRUE;
+	jboolean started = J9VMJDKINTERNALVMCONTINUATION_STARTED(currentThread, continuationObject);
+	J9VMContinuation *continuation = J9VMJDKINTERNALVMCONTINUATION_VMREF(currentThread, continuationObject);
+
+	Assert_VM_Null(currentThread->currentContinuation);
+
+	VM_ContinuationHelpers::swapFieldsWithContinuation(currentThread, continuation);
+	currentThread->currentContinuation = continuation;
+
+	Assert_VM_notNull(currentThread->currentContinuation);
+
+	if (started) {
+		/* resuming Continuation from yield */
+		resumeContinuation(currentThread, continuation);
+	} else {
+		/* start new */
+		UDATA args[] = { (UDATA) continuationObject };
+		J9NameAndSignature executeNameAndSig = { (J9UTF8*)&execute_name, (J9UTF8*)&execute_sig };
+
+		J9VMJDKINTERNALVMCONTINUATION_SET_STARTED(currentThread, continuationObject, JNI_TRUE);
+
+		runStaticMethod(currentThread, J9UTF8_DATA(&continuationClass_name), &executeNameAndSig, 1, args);
+	}
+
+	J9VMJDKINTERNALVMCONTINUATION_SET_FINISHED(currentThread, continuationObject, JNI_TRUE);
+
+	currentThread->currentContinuation = NULL;
+
+	VM_ContinuationHelpers::swapFieldsWithContinuation(currentThread, continuation);
+
+	/* For some reason the JCL swap thread objects when the VirtualThread dies, but it does
+	 * on enter and yield.
+	 */
+	currentThread->threadObject = currentThread->carrierThreadObject;
+
+	Assert_VM_Null(currentThread->currentContinuation);
+
+	return result;
+}
+
+BOOLEAN
+yieldContinuation(J9VMThread *currentThread, j9object_t scope)
+{
+	BOOLEAN result = TRUE;
+	J9VMContinuation *continuation = currentThread->currentContinuation;
+
+	/* need to check pin state before yielding */
+
+	VM_ContinuationHelpers::swapFieldsWithContinuation(currentThread, continuation);
+	/* pop the current ELS struct from the J9VMThread and store its info in J9VMContinuation struct */
+	VM_ContinuationHelpers::popAndStoreELS(currentThread, continuation);
+
+	/* Swap to parent Continuation for nested Continuation */
+	/*
+	j9object_t parentContinuation = J9VMJDKINTERNALVMCONTINUATION_PARENT(currentThread, continuationObject);
```

Comment:
i dont think we'll need to handle parent case in the VM either so this can go as well